File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/cassandra/UserDefinedTypeResultExtractorGenerator.java
package ru.tinkoff.kora.database.annotation.processor.cassandra;

import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.database.annotation.processor.entity.DbEntity;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.ArrayList;
import java.util.Objects;

public class UserDefinedTypeResultExtractorGenerator {
    private final ProcessingEnvironment processingEnv;
    private final Elements elements;
    private final Types types;

    public UserDefinedTypeResultExtractorGenerator(ProcessingEnvironment processingEnvironment) {
        this.processingEnv = processingEnvironment;
        this.elements = processingEnvironment.getElementUtils();
        this.types = processingEnvironment.getTypeUtils();
    }

    public void generate(TypeMirror type) {
        this.generateMapper(type);
        this.generateListMapper(type);
    }

    public void generateMapper(TypeMirror type) {
        var element = types.asElement(type);
        var typeName = TypeName.get(type);
        var packageName = elements.getPackageOf(element);
        var typeSpec = TypeSpec.classBuilder(CommonUtils.getOuterClassesAsPrefix(element) + element.getSimpleName() + "_CassandraRowColumnMapper")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addSuperinterface(ParameterizedTypeName.get(CassandraTypes.RESULT_COLUMN_MAPPER, typeName));
        var constructor = MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC);
        var entity = Objects.requireNonNull(DbEntity.parseEntity(this.types, type));
        this.addMappers(typeSpec, constructor, entity);

        var apply = MethodSpec.methodBuilder("apply")
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(Override.class)
            .addParameter(CassandraTypes.GETTABLE_BY_NAME, "_row")
            .addParameter(int.class, "_index")
            .returns(typeName);
        apply.addStatement("var _object = _row.getUdtValue(_index)");
        apply.beginControlFlow("if (_object == null)").addStatement("return null").endControlFlow();
        apply.addStatement("var _type = ($T) _row.getType(_index)", CassandraTypes.USER_DEFINED_TYPE);
        this.readIndexes(apply, entity);
        this.readFields(apply, entity);
        apply.addCode(entity.buildInstance("_result"));
        apply.addStatement("return _result");

        typeSpec.addMethod(apply.build());
        typeSpec.addMethod(constructor.build());

        var javaFile = JavaFile.builder(packageName.getQualifiedName().toString(), typeSpec.build()).build();
        CommonUtils.safeWriteTo(this.processingEnv, javaFile);
    }

    public void generateListMapper(TypeMirror type) {
        var element = types.asElement(type);
        var typeName = TypeName.get(type);
        var packageName = elements.getPackageOf(element);
        var typeSpec = TypeSpec.classBuilder(CommonUtils.getOuterClassesAsPrefix(element) + element.getSimpleName() + "_List_CassandraRowColumnMapper")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addSuperinterface(ParameterizedTypeName.get(CassandraTypes.RESULT_COLUMN_MAPPER, ParameterizedTypeName.get(CommonClassNames.list, typeName)));
        var constructor = MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC);
        var entity = Objects.requireNonNull(DbEntity.parseEntity(this.types, type));
        this.addMappers(typeSpec, constructor, entity);

        var apply = MethodSpec.methodBuilder("apply")
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(Override.class)
            .addParameter(CassandraTypes.GETTABLE_BY_NAME, "_row")
            .addParameter(int.class, "_index")
            .returns(ParameterizedTypeName.get(CommonClassNames.list, typeName));
        apply.addStatement("var _list = _row.getList(_index, $T.class)", CassandraTypes.UDT_VALUE);
        apply.beginControlFlow("if (_list == null)").addStatement("return null").endControlFlow();
        apply.addStatement("var _listType = ($T) _row.getType(_index)", CassandraTypes.LIST_TYPE);
        apply.addStatement("var _type = ($T) _listType.getElementType()", CassandraTypes.USER_DEFINED_TYPE);
        this.readIndexes(apply, entity);
        apply.addStatement("var _resultList = new $T<$T>(_list.size())", ArrayList.class, typeName);
        apply.beginControlFlow("for (var _object : _list)");
        this.readFields(apply, entity);
        apply.addCode(entity.buildInstance("_result"));
        apply.addStatement("_resultList.add(_result)");
        apply.endControlFlow();
        apply.addStatement("return _resultList");

        typeSpec.addMethod(apply.build());
        typeSpec.addMethod(constructor.build());

        var javaFile = JavaFile.builder(packageName.getQualifiedName().toString(), typeSpec.build()).build();
        CommonUtils.safeWriteTo(this.processingEnv, javaFile);
    }


    private void readFields(MethodSpec.Builder apply, DbEntity entity) {
        for (var entityField : entity.columns()) {
            var fieldName = entityField.element().getSimpleName().toString();
            var index = CodeBlock.of("$N", "_index_of_" + entityField.element().getSimpleName());
            var nativeType = CassandraNativeTypes.findNativeType(TypeName.get(entityField.type()));
            if (nativeType != null) {
                apply.addStatement("var $N = $L", fieldName, nativeType.extract("_object", index));
            } else {
                var mapperName = "_" + fieldName + "_mapper";
                apply.addStatement("var $N = this.$N.apply(_object, $L)", fieldName, mapperName, index);
            }
        }
        apply.addCode("\n");
    }

    private void addMappers(TypeSpec.Builder typeSpec, MethodSpec.Builder constructor, DbEntity entity) {
        for (var entityField : entity.columns()) {
            var nativeType = CassandraNativeTypes.findNativeType(TypeName.get(entityField.type()));
            if (nativeType == null) {
                var mapperName = "_" + entityField.element().getSimpleName() + "_mapper";
                // todo mapping annotation support?
                var mapperType = ParameterizedTypeName.get(CassandraTypes.RESULT_COLUMN_MAPPER, TypeName.get(entityField.type()));
                constructor.addParameter(mapperType, mapperName);
                constructor.addStatement("this.$N = $N", mapperName, mapperName);
                typeSpec.addField(mapperType, mapperName, Modifier.PRIVATE, Modifier.FINAL);
            }
        }
    }

    private void readIndexes(MethodSpec.Builder apply, DbEntity entity) {
        for (var entityField : entity.columns()) {
            apply.addStatement("var $N = _type.firstIndexOf($S)", "_index_of_" + entityField.element().getSimpleName(), entityField.columnName());
        }
        apply.addCode("\n");
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/cassandra/udt/UserDefinedTypeResultExtractorGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.cassandra.udt

import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import com.squareup.kotlinpoet.ksp.toTypeName
import com.squareup.kotlinpoet.ksp.writeTo
import ru.tinkoff.kora.database.symbol.processor.cassandra.CassandraNativeTypes
import ru.tinkoff.kora.database.symbol.processor.cassandra.CassandraTypes
import ru.tinkoff.kora.database.symbol.processor.model.DbEntity
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.generatedClassName

class UserDefinedTypeResultExtractorGenerator(private val environment: SymbolProcessorEnvironment) {
    fun generate(resolver: Resolver, classDeclaration: KSClassDeclaration) {
        this.generateRowColumnMapper(resolver, classDeclaration)
        this.generateListRowColumnMapper(resolver, classDeclaration)
    }

    private fun generateRowColumnMapper(resolver: Resolver, classDeclaration: KSClassDeclaration) {
        val type = classDeclaration.asType(listOf())
        val typeName = type.toTypeName().copy(false)
        val typeSpec = TypeSpec.classBuilder(classDeclaration.generatedClassName("CassandraRowColumnMapper"))
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
            .addSuperinterface(CassandraTypes.rowColumnMapper.parameterizedBy(typeName))
        classDeclaration.containingFile?.let { typeSpec.addOriginatingKSFile(it) }
        val constructor = FunSpec.constructorBuilder()
        val entity = DbEntity.parseEntity(type)!!
        this.addMappers(typeSpec, constructor, entity)

        val apply = FunSpec.builder("apply")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("_row", CassandraTypes.gettableByName)
            .addParameter("_index", Int::class.javaPrimitiveType!!.asTypeName())
            .returns(typeName.copy(true))
        apply.addStatement("val _type = _row.getType(_index) as %T", CassandraTypes.userDefinedType)
        apply.addStatement("val _object = _row.getUdtValue(_index)")
        apply.controlFlow("if (_object == null)") {
            addStatement("return null")
        }
        apply.addCode("\n")
        this.readIndexes(apply, entity)
        this.readFields(apply, entity)
        apply.addCode("return %L\n", buildEntity(entity))
        typeSpec.addFunction(apply.build())
        typeSpec.primaryConstructor(constructor.build())

        FileSpec.get(classDeclaration.packageName.asString(), typeSpec.build()).writeTo(environment.codeGenerator, false, listOfNotNull(classDeclaration.containingFile))
    }

    private fun generateListRowColumnMapper(resolver: Resolver, classDeclaration: KSClassDeclaration) {
        val type = classDeclaration.asType(listOf())
        val typeName = type.toTypeName().copy(false)
        val listTypeName = LIST.parameterizedBy(typeName).copy(false)
        val typeSpec = TypeSpec.classBuilder(classDeclaration.generatedClassName("List_CassandraRowColumnMapper"))
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
            .addSuperinterface(CassandraTypes.rowColumnMapper.parameterizedBy(listTypeName))
        classDeclaration.containingFile?.let { typeSpec.addOriginatingKSFile(it) }
        val constructor = FunSpec.constructorBuilder()
        val entity = DbEntity.parseEntity(type)!!
        this.addMappers(typeSpec, constructor, entity)

        val apply = FunSpec.builder("apply")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("_row", CassandraTypes.gettableByName)
            .addParameter("_index", Int::class.javaPrimitiveType!!.asTypeName())
            .returns(listTypeName.copy(true))
        apply.addStatement("val _listType = _row.getType(_index) as %T", CassandraTypes.listType)
        apply.addStatement("val _type = _listType.getElementType() as %T", CassandraTypes.userDefinedType)
        apply.addStatement("val _list = _row.getList(_index, %T::class.java)", CassandraTypes.udtValue)
        apply.controlFlow("if (_list == null)") {
            addStatement("return null")
        }
        readIndexes(apply, entity)
        apply.addCode("\n")
        apply.controlFlow("return _list.map { _object ->") {
            readFields(this, entity)
            apply.addCode("%L\n", buildEntity(entity))
        }

        typeSpec.addFunction(apply.build())
        typeSpec.primaryConstructor(constructor.build())

        FileSpec.get(classDeclaration.packageName.asString(), typeSpec.build()).writeTo(environment.codeGenerator, false, listOfNotNull(classDeclaration.containingFile))
    }

    private fun readIndexes(apply: FunSpec.Builder, entity: DbEntity) {
        for (entityField in entity.columns) {
            val fieldName = entityField.property.simpleName.asString()
            apply.addStatement("val %N = _type.firstIndexOf(%S)", "_index_of_$fieldName", entityField.columnName)
        }
        apply.addCode("\n")
    }

    private fun readFields(apply: FunSpec.Builder, entity: DbEntity) {
        for (entityField in entity.columns) {
            val fieldName = entityField.property.simpleName.asString()
            val fieldTypeName = entityField.type.toTypeName()
            val nativeType = CassandraNativeTypes.findNativeType(fieldTypeName)
            apply.controlFlow("val %N = if (_object.isNull(%N))", fieldName, "_index_of_$fieldName") {
                if (fieldTypeName.isNullable) {
                    addStatement("null")
                } else {
                    addStatement("throw %T(%S)", NullPointerException::class, "Field $fieldName is not nullable, but column ${entityField.columnName} is null")
                }
                nextControlFlow("else")
                if (nativeType != null) {
                    addCode(nativeType.extract("_object", CodeBlock.of("%N", "_index_of_$fieldName")))
                } else {
                    val mapperName = "_${fieldName}_mapper"
                    addCode("this.%N.apply(_object, %N)", mapperName, "_index_of_$fieldName")
                }
                addCode("!!\n")
            }
        }
    }

    private fun addMappers(typeSpec: TypeSpec.Builder, constructor: FunSpec.Builder, entity: DbEntity) {
        for (entityField in entity.fields) {
            val fieldName = entityField.property.simpleName.asString()
            val fieldTypeName = entityField.type.toTypeName()
            val nativeType = CassandraNativeTypes.findNativeType(fieldTypeName)
            if (nativeType != null) {
                continue
            }
            val mapperName = "_${fieldName}_mapper"
            val mapperType = CassandraTypes.rowColumnMapper.parameterizedBy(fieldTypeName.copy(false))
            constructor.addParameter(mapperName, mapperType)
            constructor.addStatement("this.%N = %N", mapperName, mapperName)

            typeSpec.addProperty(mapperName, mapperType, KModifier.PRIVATE, KModifier.FINAL)
        }
    }

    private fun buildEntity(entity: DbEntity): CodeBlock {
        val b = CodeBlock.builder()
        b.add("%T(\n", entity.type.toTypeName())
        for (i in 0 until entity.fields.size) {
            if (i > 0) {
                b.add(",\n")
            }
            val field = entity.fields[i]
            val fieldName = field.property.simpleName.asString()
            b.add("  %N", fieldName)
        }
        b.add("\n)")
        return b.build()
    }
}
